README.md
# log-defer

![Unit Tests](https://github.com/ioncache/log-defer/actions/workflows/nodejs.yml/badge.svg) [![codecov](https://codecov.io/gh/ioncache/log-defer/branch/main/graph/badge.svg)](https://codecov.io/gh/ioncache/log-defer)

## Description

Generate log object compatible with [log-defer-viz](https://github.com/hoytech/Log-Defer-Viz)

log-defer is a module that creates structured logs. The log-defer documentation explains structured logging and its benefits over ad-hoc logging.

This is the javascript implementation, full documentation for log-defer can be found at <https://metacpan.org/pod/Log::Defer>

## Installation

``` bash
npm install log-defer
```

or

```bash
bower install log-defer
```

## Synopsis

```javascript
'use strict';

var log = require('log-defer');
var process = require('process');

log.timer('Log Messages'); // begin a timer called 'Log Messages'
log.info('Info level logging');
log.warn('Warn level logging');
log.error('Error level logging');
log.debug('Debug level logging');
log.timer('Log Messages'); // end the 'Log Messages' timer
log.data({ bar: 'baz', foo: 'bar' }); // add to the data object in the log-defer

log.warn('Warning with data!', { bar: 'baz', foo: 'bar' }, { barbaz: 'bazfoo', foobar: 'barbaz' });

var output = log.finalizeLog(); // finalize the log, and return a json string of the log-defer

process.stdout.write(output);
```

The above code should produce the following output:

```bash
{"data":{"bar":"baz","foo":"bar"},"logs":[[0.0009999275207519531,30,"Info level logging"],[0.0009999275207519531,20,"Warn level logging"],[0.0009999275207519531,10,"Error level logging"],[0.0009999275207519531,40,"Debug level logging"],[0.0009999275207519531,20,"Warning with data!",{"bar":"baz","foo":"bar","barbaz":"bazfoo","foobar":"barbaz"}]],"start":1457018806.914,"timers":[["Log Messages",0.0009999275207519531,0.0009999275207519531]],"end":0.0019998550415039062}
```

If viewed using the log-defer-viz cli then it would format as follows:

```bash
------ 2016-03-03 Thu 10:26:46.914 EST (1457018806.914) ------
  | 0.001000 [ INFO] Info level logging
  | 0.001000 [ WARN] Warn level logging
  | 0.001000 [ERROR] Error level logging
  | 0.001000 [DEBUG] Debug level logging
  | 0.001000 [ WARN] Warning with data! [{"foo":"bar","bar":"baz","foobar":"barbaz","barbaz":"bazfoo"}]
  |_0.002000 [END]

 Log Messages                                                                                  X
_______________________________________________________________________________________________________
times in ms                                                                                    1.0

  Data:
{
   "bar" : "baz",
   "foo" : "bar"
}
```


CHANGES.md
# log-defer Change Log

## 1.0.0 - 2021-06-10

Updated:

- numerous dependency updates
- removed node 10.x from automated tests
  - due to dev dependency updates, the unit tests no longer pass in node 10.x, but the code itself has not changed so it should still run

## 0.0.8 - 2020-07-09

Updated:

- dependency updates

## 0.0.7 - 2016-03-04

New:

- added html demo

## 0.0.6 - 2016-03-03

New:

- allow for amd and browser usage

Updated:

- updates to demos

Fixed:

- properly sort timers by start time

## 0.0.5 - 2016-03-03

New:

- sort timers to ensure in ascending order by time
- added README.md
- added CHANGES.md

## 0.0.1 - 2016-03-03

New:

- initial release


test/tests.js
'use strict';

var _ = require('lodash');
var chai = require('chai');
var log = require('../lib/log-defer');
var q = require('q');

var expect = chai.expect;

afterEach(function () {
  log.finalizeLog();
});

describe("function 'addMessage':", function () {
  it('should add a message to the current log list', function () {
    // Arrange
    var message = 'message adding test';

    // Act
    log.addMessage(message);

    // Assert
    expect(log._self.message.logs.length).to.equal(1);
    expect(log._self.message.logs[0][2]).to.equal(message);
  });

  describe('when there is a level option', function () {
    it('should set the level to the correct value', function () {
      // Arrange
      var message = 'setting level test';

      // Act
      log.addMessage(message, 10);

      // Assert
      expect(log._self.message.logs[0][1]).to.equal(10);
    });
  });

  describe('when there is a data option', function () {
    it('should combine all into the log', function () {
      // Arrange
      var data = { foo: 'bar' };
      var data2 = { bar: 'baz' };
      var data3 = { baz: 'meh' };
      var message = 'setting data';

      // Act
      log.addMessage(message, data, data2, data3);

      // Assert
      expect(log._self.message.logs[0][3]).to.eql(_.extend({}, data, data2, data3));
    });

    // not sure why non-object args are ignored
    // would need to go back to the log-defer spec
    it('should ignore non-object arguments', function () {
      // Arrange
      var data = { foo: 'bar' };
      var data2 = { bar: 'baz' };
      var data3 = [
        {
          baz: 'meh',
          qux: {
            quux: 1,
            quuz: 2,
            corge: '4',
          },
        },
      ];
      var data4 = 0;
      var data5 = 'this is not an object';
      var message = 'setting data';

      // Act
      log.addMessage(message, data, data2, data3, data4, data5);

      // Assert
      expect(log._self.message.logs[0][3]).to.eql(_.extend({}, data, data2));
    });
  });

  describe('when there is an invalid log message', function () {
    it('should do nothing', function () {
      // Arrange
      var data = { foo: 'bar' };
      var data2 = { bar: 'baz' };
      var data3 = { baz: 'meh' };
      var message = 'setting data';

      // Act
      log.addMessage(0, data, data2, data3);

      // Assert
      expect(log._self.message.logs.length).to.eql(0);

      // Act
      log.addMessage();

      // Assert
      expect(log._self.message.logs.length).to.eql(0);
    });
  });
});

describe("function 'timer':", function () {
  describe('when called and the timer name does not exist', function () {
    it('then a new timer should be started', function () {
      // Arrange
      var timerName = 'foo';

      // Act
      log.timer(timerName);

      // Assert
      expect(log._self.message.timers[timerName]).to.exist;
      expect(log._self.message.timers[timerName].start).to.exist;
      expect(log._self.message.timers[timerName].end).to.not.exist;
    });
  });

  describe('when called and the timer name does exist', function () {
    it('then an end time should be added to the timer', function () {
      // Arrange
      var timerName = 'foo';

      // Act
      log.timer(timerName);
      log.timer(timerName);

      // Assert
      expect(log._self.message.timers[timerName].end).to.exist;
    });
  });
});

describe("function 'finalizeLog':", function () {
  it('should return a json string', function () {
    // Act
    log.info('this is a test');
    var output = log.finalizeLog();

    // Assert
    expect(typeof output).to.equal('string');
    expect(typeof JSON.parse(output)).to.equal('object');
  });

  it('should clear the current log', function () {
    // Act
    log.info('this is a test');
    log.finalizeLog();

    // Assert
    expect(log._self.message.data).to.be.empty;
    expect(log._self.message.logs).to.be.empty;
    expect(log._self.message.timers).to.be.empty;
  });
});

describe('Log level helpers', function () {
  describe('error', function () {
    it('should set the log level to 10', function () {
      // Act
      log.error('this is an error test');
      var outputObj = JSON.parse(log.finalizeLog());

      // Assert
      expect(outputObj.logs[0][1]).to.equal(10);
    });
  });

  describe('warn', function () {
    it('should set the log level to 20', function () {
      // Act
      log.warn('this is a warn test');
      var outputObj = JSON.parse(log.finalizeLog());

      // Assert
      expect(outputObj.logs[0][1]).to.equal(20);
    });
  });

  describe('info', function () {
    it('should set the log level to 30', function () {
      // Act
      log.info('this is an info test');
      var outputObj = JSON.parse(log.finalizeLog());

      // Assert
      expect(outputObj.logs[0][1]).to.equal(30);
    });
  });

  describe('debug', function () {
    it('should set the log level to 40', function () {
      // Act
      log.debug('this is a debug');
      var outputObj = JSON.parse(log.finalizeLog());

      // Assert
      expect(outputObj.logs[0][1]).to.equal(40);
    });
  });
});

describe('multiple logs, timers, and data', function () {
  it('the output should match all log statements', function () {
    // Act
    var deferreds = [];

    log.timer('Log Messaging Demo'); // begin a timer called 'Log Messaging Demo'

    log.info('Info level logging');
    log.warn('Warn level logging');
    log.error('Error level logging');
    log.timer('Interleaved, longer running timer');
    log.debug('Debug level logging');

    log.timer('async timer');
    log.info('starting async timer');
    var deferred1 = q.defer();
    deferreds.push(deferred1.promise);
    setTimeout(function () {
      log.info('finishing async timer');
      log.timer('async timer');
      deferred1.resolve();
    }, 375);

    log.timer('Short timer');
    var deferred2 = q.defer();
    deferreds.push(deferred2);
    setTimeout(function () {
      log.timer('Short timer');
      deferred2.resolve();
    }, 125);

    log.data({ bar: 'baz', foo: 'bar' }); // add to the data object in the log-defer
    log.warn('Warning with data!', { bar: 'baz', foo: 'bar' }, { barbaz: 'bazfoo', foobar: 'barbaz' });

    var deferred3 = q.defer();
    deferreds.push(deferred3.promise);
    setTimeout(function () {
      log.timer('Interleaved, longer running timer');
      deferred3.resolve();
    }, 700);

    log.data({ quux: 'xyzzy', qux: 'corge' });

    // Assert
    q.all(deferreds).then(function () {
      log.timer('Log Messaging Demo'); // end the 'Log Messaging Demo' timer

      var output = log.finalizeLog(); // finalize the log, and return a json string of the log-defer
      var outputObj = JSON.parse(output);

      expect(outputObj).to.be.defined;
    });
  });
});


tests/test.js
'use strict';

var log = require('../lib/log-defer');

console.log(log);

log.error('this is a test');

var output = log.finalizeLog();

console.log(output);


lib/index.js
(function() {

  // Ported from python version https://github.com/mikep/LogDefer/blob/master/LogDefer.py

  var _this = this;

  _this.levels = [40, 30, 20, 10];
  _this.message = {
    data: {},
    logs: [],
    start: new Date().getTime(),
    timers: {},
  };

  // add message to log object
  _this.addMessage = function(params) {
    var data = params.data || {};
    var level = params.level || 30;
    var message = params.message || "";
    
    var log = [_this._getEt(), level, message];

    if (data && arguments.length > 1) {
      for (var i = 1; i < arguments.length; i++ ) {
        Object.assign(data, arguments[i]);
      }

      log.append(data);
    }

    _this.message.logs.append(log);
  };

  // add timer to log object
  // if timer already exists, set the end time
  _this.timer = function(name) {
    _this.name = name;

    if (name && !_this.message.timers.hasOwnProperty(name)) {
      _this.message.timers[name] = {
        name: name,
        start: _this._getEt(),
      };
    } else {
      _this.message.timers[name].end = _this._getEt();
    }
  };

  // log-defer-viz uses time since the start time in logs and timers
  _this._getEt = function() {
    return (new Date().getTime()) - _this.message.start;
  };

  return function(params) {
    // TODO
  };

}());


lib/log-defer.js
/*
  Generate log object compatible with log-defer-viz
  https://github.com/hoytech/Log-Defer-Viz

  Ported from python version https://github.com/mikep/LogDefer/blob/master/LogDefer.py
*/

(function (global) {
  'use strict';

  var self = {
    message: {
      data: {},
      logs: [],
      start: new Date().getTime() / 1000,
      timers: {},
    },
  };

  /*
    Public methods
    ==============
  */

  // add message to log object
  self.addMessage = function () {
    if (arguments.length < 0 || typeof arguments[0] !== 'string') {
      return;
    }

    var args = Array.prototype.slice.call(arguments);

    var data = {};
    var level = 30;
    var message = args.shift();

    if (typeof args[0] === 'number') {
      level = args.shift();
    }

    var log = [self._getEt(), level, message];

    if (args.length > 0) {
      for (var i = 0; i < args.length; i++) {
        var arg = args[i];
        if (typeof arg === 'object' && !Array.isArray(arg)) {
          Object.assign(data, arg);
        }
      }
    }

    if (Object.keys(data).length) {
      log.push(data);
    }

    self.message.logs.push(log);
  };

  // add timer to log object
  // if timer already exists, set the end time
  self.timer = function (name) {
    self.name = name;

    if (name && !self.message.timers.hasOwnProperty(name)) {
      self.message.timers[name] = {
        name: name,
        start: self._getEt(),
      };
    } else if (self.message.timers.hasOwnProperty(name)) {
      self.message.timers[name].end = self._getEt();
    }
  };

  self.data = function (data) {
    if (typeof data === 'object' && !Array.isArray(data)) {
      Object.assign(self.message.data, data);
    }
  };

  // format and return the log object for logging
  self.finalizeLog = function () {
    self._formatLogMessageOutput();

    var output = JSON.stringify(self.message);

    // reset after logging
    self.message.data = {};
    self.message.logs = [];
    self.message.timers = {};

    return output;
  };

  // log level functions
  self.debug = function () {
    var args = Array.prototype.slice.call(arguments);
    args.splice(1, 0, 40);
    self.addMessage.apply(this, args);
  };

  self.error = function () {
    var args = Array.prototype.slice.call(arguments);
    args.splice(1, 0, 10);
    self.addMessage.apply(this, args);
  };

  self.info = function () {
    var args = Array.prototype.slice.call(arguments);
    args.splice(1, 0, 30);
    self.addMessage.apply(this, args);
  };

  self.warn = function () {
    var args = Array.prototype.slice.call(arguments);
    args.splice(1, 0, 20);
    self.addMessage.apply(this, args);
  };

  /*
    Private methods
    ===============
  */

  self._formatLogMessageOutput = function () {
    // clean up, log-defer-viz doesn't like empty objects
    ['logs', 'timers', 'data'].forEach(function (key) {
      if (
        self.message.hasOwnProperty(key) &&
        ((self.message[key].hasOwnProperty('length') && self.message[key].length === 0) ||
          Object.keys(self.message[key]).length === 0)
      ) {
        delete self.message[key];
      }
    });

    // convert timers to a list
    if (self.message.hasOwnProperty('timers')) {
      var timers = [];
      Object.keys(self.message.timers).forEach(function (key) {
        timers.push([
          self.message.timers[key].name,
          self.message.timers[key].start,
          self.message.timers[key].end || self._getEt(),
        ]);
      });

      timers.sort(function (a, b) {
        return a[1] >= b[1];
      });

      self.message.timers = timers;
    }

    // record end time
    self.message.end = self._getEt();
  };

  // log-defer-viz uses time since the start time in logs and timers
  self._getEt = function () {
    return new Date().getTime() / 1000 - self.message.start;
  };

  var logDefer = {
    _self: self, // for testing only
    addMessage: self.addMessage,
    data: self.data,
    debug: self.debug,
    error: self.error,
    finalizeLog: self.finalizeLog,
    info: self.info,
    timer: self.timer,
    warn: self.warn,
  };

  // no strong need to test the import code
  /* istanbul ignore next */
  if (typeof define === 'function' && define.amd) {
    define('log-defer', [], logDefer);
  } else if (typeof module === 'object' && module.exports) {
    module.exports = logDefer;
  } else {
    global.logDefer = logDefer;
  }
})(this);


demo/demo.js
'use strict';

// npm modules
var log = require('../lib/log-defer');
var process = require('process');
var q = require('q');

var deferreds = [];

log.timer('Log Messaging Demo'); // begin a timer called 'Log Messaging Demo'

log.info('Info level logging');
log.warn('Warn level logging');
log.error('Error level logging');
log.timer('Interleaved, longer running timer');
log.debug('Debug level logging');

log.timer('async timer');
log.info('starting async timer');
var deferred1 = q.defer();
deferreds.push(deferred1.promise);
setTimeout(function() {
  log.info('finishing async timer');
  log.timer('async timer');
  deferred1.resolve();
}, 375);

log.timer('Short timer');
var deferred2 = q.defer();
deferreds.push(deferred2);
setTimeout(function() {
  log.timer('Short timer');
  deferred2.resolve();
}, 125);

log.data({ bar: 'baz', foo: 'bar' }); // add to the data object in the log-defer
log.warn('Warning with data!', { bar: 'baz', foo: 'bar' }, { barbaz: 'bazfoo', foobar: 'barbaz' });

var deferred3 = q.defer();
deferreds.push(deferred3.promise);
setTimeout(function() {
  log.timer('Interleaved, longer running timer');
  deferred3.resolve();
}, 700);

log.data({ quux: 'xyzzy', qux: 'corge' });

// add some delay in so entire run time is at least a second
var deferred4 = q.defer();
deferreds.push(deferred4.promise);
setTimeout(function() {
  deferred4.resolve();
}, 1000);

q.all(deferreds).then(function() {
  log.timer('Log Messaging Demo'); // end the 'Log Messaging Demo' timer
  var output = log.finalizeLog(); // finalize the log, and return a json string of the log-defer

  process.stdout.write(output);
});
